dolfin_adjoint_common/blocks/constant.py

```diff
@@ -24,16 +26,22 @@ class ConstantAssignBlock(Block):
     def __init__(self, other):
         super(ConstantAssignBlock, self).__init__()
         self.add_dependency(other)
+        self.assigned_float = isinstance(other, float)
 
     def evaluate_adj_component(self, inputs, adj_inputs, block_variable, idx, prepared=None):
-        return adj_inputs[0]
+        adj_output = adj_inputs[0]
+        if self.assigned_float:
+            # Convert to float
+            adj_output = gather(adj_output)
+            adj_output = float(adj_output)
+        return adj_output
 
     def evaluate_tlm_component(self, inputs, tlm_inputs, block_variable, idx, prepared=None):
-        return tlm_inputs[0]
+        return constant_from_values(block_variable.output, tlm_inputs[0])
```

Comment:
This is the same for Functions as well. It is mostly because that was easier in the early stages of development: assembling the derivative form - i.e a 1-form, returned a vector in FEniCS, and multiplying vectors instead of Functions with the adj_input worked better since the latter becomes a UFL expression.

Initially the tlm_value also used vector types, but we came to the realization that during forward AD it is much easier and natural to use the same type (At least for the methods we were overloading).

That said, I am of course open to changing it if it simplifies things.

It is sort of briefly mentioned in the pyadjoint documentation, but perhaps not in a very clear way:
http://www.dolfin-adjoint.org/en/latest/documentation/pyadjoint_docs.html#blockvariable